lib/bolt_server/transport_app.rb

```diff
@@ -142,6 +139,84 @@ def run_task(target, body)
       end
     end
 
+    def run_task(target, body)
+      validate_schema(@schemas["action-run_task"], body)
+
+      task_data = body['task']
+      task = Bolt::Task::PuppetServer.new(task_data['name'], task_data['metadata'], task_data['files'], @file_cache)
+      task_helper(target, task, body['parameters'] || {})
+    end
+
+    def extract_install_task(target)
+      unless target.plugin_hooks['puppet_library']['task']
+        raise BoltServer::RequestError,
+              "Target must have 'task' plugin hook"
+      end
+      install_task = target.plugin_hooks['puppet_library']['task'].split('::', 2)
+      install_task << 'init' if install_task.count == 1
+      install_task
+    end
+
+    def apply_prep(target, body)
+      validate_schema(@schemas["action-apply_prep"], body)
+      plugins_tarball = if (tarball = @file_cache.get_cached_plugin_tarball(body['versioned_project'], 'fact_plugins'))
```

Comment:
Since this logic is also duplicated in `apply`, I think it would be good to add a `build_cached_project_plugins_tarball`-esque helper that centralizes this stuff.

Reply:
Also, I think the code here risks running into weird "simultaneous file-write" edge-cases resulting from at least two threads trying to write to the cache. Fortunately, file_cache has a nice pattern for that via serial_execute (see e.g. https://github.com/puppetlabs/bolt/blob/main/lib/bolt_server/transport_app.rb#L182). Reading the linked example, I think wrapping all of this plugin-tarball-computation code up in a `file_cache.serial_execute` should handle that edge-case (including the code that reads from the cache).

^ is also further motivation for the centralized `build_cached_project_plugins_tarball` helper.

Reply:
oh jeeze i thought ruby had atomic writes. yeah i for sure need to fix this up. Thanks

Reply:
Updated and added tests. Verified that the tests fail without the serial_execute. Great catch Enis!